File: src/FindFirstCommonNode.cpp
/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution {
public:
    ListNode* FindFirstCommonNode(ListNode *pHead1, ListNode *pHead2) {
        int len1 = GetListLength(pHead1);
        int len2 = GetListLength(pHead2);

        int diff = abs(len1 - len2);

        ListNode *pLongHead, *pShortHead;
        if (len1 > len2) {
          pLongHead = pHead1;
          pShortHead = pHead2;
        } else {
          pLongHead = pHead2;
          pShortHead = pHead1;
        }

        while (diff > 0 && pLongHead != NULL) {
          pLongHead = pLongHead->next;
          diff--;
        }

        while (pLongHead != NULL && pShortHead != NULL && pLongHead != pShortHead) {
          pLongHead = pLongHead->next;
          pShortHead = pShortHead->next;
        }

        return pLongHead;
    }

    int GetListLength(ListNode *head) {
      if (head == NULL) {
        return 0;
      }
      ListNode *p = head;
      int len = 0;
      while (p != NULL) {
        p = p->next;
        len++;
      }
      return len;
    }
};

File: src/java/FindFirstCommonNode.java
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode FindFirstCommonNode(ListNode pHead1, ListNode pHead2) {
        if (pHead1 == null || pHead2 == null) return null;

        ListNode p1 = pHead1;
        ListNode p2 = pHead2;
        int n = 0;
        while (p1.next != null) {
            p1 = p1.next;
            n++;
        }
        while (p2.next != null) {
            p2 = p2.next;
            n--;
        }

        if (p1 != p2) {
            return null;
        }

        p1 = n > 0 ? pHead1 : pHead2;	// longer list
        p2 = p1 == pHead1 ? pHead2 : pHead1;	// shorter list
        n = Math.abs(n);
        while (n != 0) {
            p1 = p1.next;
            n--;
        }
        while (p1 != p2) {
            p1 = p1.next;
            p2 = p2.next;
        }

        return p1;
    }
}
